mautic/mautic
# Pipedrive activities support

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | Y
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
PR added sync Mautic activities (Email read, Page hit, Change segments membership...) to Pipedrive.
This PR added new options to Pipedrive plugin settings and new command 
`php app/console  mautic:integration:pipedrive:pushactivities`

Description:
- If activitiy not exists (Email read...), plugin will create new one https://developers.pipedrive.com/docs/api/v1/#!/ActivityTypes
- Then sync activities to contact profile https://developers.pipedrive.com/docs/api/v1/#!/Activities/post_activities
- If event name is clickable in Mautic (email, page visits), then  will add clickable link to activity description 
![image](https://user-images.githubusercontent.com/462477/52562275-4ee3b600-2dfe-11e9-96c6-11c0247b2922.png)
- sync by command/cron  `php app/console  mautic:integration:pipedrive:pushactivities`  - this command use time-interval parameter (default 15 minutes)  - prevents sync activities to much often
- plugin store **last_activity_sync** timestamp to integration_entity table to column internal. If **last_activity_sync**  doesn't exists, then sync all events , If exists then just sync events from last_activity_sync to now.

![image](https://user-images.githubusercontent.com/462477/52562360-923e2480-2dfe-11e9-8a2d-a512d4553132.png)

This PR also improves batch processing for import contacts to Pipedrive: 
https://github.com/mautic/mautic/pull/7233/files#diff-b03bbd4c72ad0f3079918af2392a8fd7R92


#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com) and setup Pipedrive plugin https://www.mautic.org/docs/en/plugins/pipedrive.html
2. setup sync activities  to Pipedrive
![image](https://user-images.githubusercontent.com/462477/52562074-974ea400-2dfd-11e9-9336-358d9696e6dd.png)
3. sync contacts to Pipedrive `php app/console mautic:integration:pipedrive:push`
4. sync activities to Pipedrive `php app/console  mautic:integration:pipedrive:pushactivities`
5. Go to Pipedrive and see activities in each contacts
![image](https://user-images.githubusercontent.com/462477/52562128-bbaa8080-2dfd-11e9-8d4c-d8fbe285da79.png)
